File: backend/suggestions.py
"""
GazeTalk Backend — Word suggestion / prediction engine.
Uses a frequency-ranked English word list with prefix matching.
Returns top-N completions for a partial word input.
"""
import logging
from typing import List

logger = logging.getLogger("gztalk.suggestions")

# Top ~500 most common English words, frequency-ordered
# This is a lightweight, offline approach — no external deps needed
COMMON_WORDS = [
    "the", "be", "to", "of", "and", "a", "in", "that", "have", "i",
    "it", "for", "not", "on", "with", "he", "as", "you", "do", "at",
    "this", "but", "his", "by", "from", "they", "we", "say", "her", "she",
    "or", "an", "will", "my", "one", "all", "would", "there", "their", "what",
    "so", "up", "out", "if", "about", "who", "get", "which", "go", "me",
    "when", "make", "can", "like", "time", "no", "just", "him", "know", "take",
    "people", "into", "year", "your", "good", "some", "could", "them", "see", "other",
    "than", "then", "now", "look", "only", "come", "its", "over", "think", "also",
    "back", "after", "use", "two", "how", "our", "work", "first", "well", "way",
    "even", "new", "want", "because", "any", "these", "give", "day", "most", "us",
    "great", "help", "need", "house", "water", "food", "please", "thank", "thanks",
    "hello", "hi", "goodbye", "bye", "yes", "no", "okay", "sorry", "excuse",
    "stop", "start", "go", "wait", "come", "here", "there", "left", "right",
    "up", "down", "more", "less", "big", "small", "hot", "cold", "good", "bad",
    "happy", "sad", "hungry", "thirsty", "tired", "sick", "pain", "hurt",
    "doctor", "hospital", "medicine", "emergency", "call", "phone", "home",
    "family", "friend", "mother", "father", "brother", "sister", "child",
    "love", "feel", "want", "need", "can", "eat", "drink", "sleep", "walk",
    "sit", "stand", "open", "close", "turn", "light", "dark", "room", "door",
    "bathroom", "kitchen", "bed", "chair", "table", "outside", "inside",
    "morning", "afternoon", "evening", "night", "today", "tomorrow", "yesterday",
    "what", "where", "when", "why", "how", "much", "many", "again", "always",
    "never", "sometimes", "soon", "later", "before", "after", "now",
    "very", "really", "too", "enough", "just", "still", "already", "almost",
    "understand", "speak", "listen", "read", "write", "learn", "teach",
    "name", "age", "address", "number", "school", "job", "money", "price",
    "beautiful", "wonderful", "amazing", "important", "different", "same",
    "together", "alone", "quiet", "loud", "fast", "slow", "easy", "hard",
]

# Deduplicate while preserving order
_seen = set()
WORD_LIST: List[str] = []
for w in COMMON_WORDS:
    if w not in _seen:
        _seen.add(w)
        WORD_LIST.append(w)


def suggest(partial: str, limit: int = 5) -> List[str]:
    """Return up to `limit` word completions for a partial string.
    Prioritises frequency rank (earlier in list = more common).
    """
    if not partial or not partial.strip():
        return []

    prefix = partial.strip().lower()
    results = []

    for word in WORD_LIST:
        if word.startswith(prefix) and word != prefix:
            results.append(word)
            if len(results) >= limit:
                break

    return results


def suggest_next_word(context: str, limit: int = 5) -> List[str]:
    """Given the full typed text so far, suggest the next most likely word.
    Currently uses simple heuristics — could be replaced with an ML model.
    """
    if not context or not context.strip():
        # Common conversation starters
        return ["hello", "i", "please", "help", "thank"]

    words = context.strip().lower().split()
    last_word = words[-1] if words else ""

    # If the last word looks incomplete (no space at end), suggest completions
    if not context.endswith(' '):
        return suggest(last_word, limit)

    # Otherwise suggest common follow-up words
    FOLLOW_UPS = {
        "i": ["want", "need", "am", "feel", "can"],
        "please": ["help", "call", "give", "open", "come"],
        "can": ["you", "i", "we", "help", "go"],
        "help": ["me", "please", "now", "call", "emergency"],
        "want": ["to", "food", "water", "help", "go"],
        "need": ["help", "water", "food", "medicine", "doctor"],
        "thank": ["you", "thanks", "so", "very", "much"],
        "go": ["home", "outside", "to", "now", "there"],
        "feel": ["good", "bad", "sick", "tired", "hungry"],
        "am": ["good", "hungry", "tired", "sick", "happy"],
    }

    suggestions = FOLLOW_UPS.get(last_word, [])
    if suggestions:
        return suggestions[:limit]

    # Default: common words
    return ["please", "help", "want", "need", "thank"][:limit]

File: src/utils/suggestions.ts
/**
 * GazeTalk — Client-side word suggestion engine.
 * Runs entirely on-device — no backend required.
 * Ported from backend/suggestions.py with identical logic.
 */

// Top ~200 most common English words, frequency-ordered.
// Curated for AAC (Augmentative & Alternative Communication) use cases.
const COMMON_WORDS: string[] = [
  'the', 'be', 'to', 'of', 'and', 'a', 'in', 'that', 'have', 'i',
  'it', 'for', 'not', 'on', 'with', 'he', 'as', 'you', 'do', 'at',
  'this', 'but', 'his', 'by', 'from', 'they', 'we', 'say', 'her', 'she',
  'or', 'an', 'will', 'my', 'one', 'all', 'would', 'there', 'their', 'what',
  'so', 'up', 'out', 'if', 'about', 'who', 'get', 'which', 'go', 'me',
  'when', 'make', 'can', 'like', 'time', 'no', 'just', 'him', 'know', 'take',
  'people', 'into', 'year', 'your', 'good', 'some', 'could', 'them', 'see', 'other',
  'than', 'then', 'now', 'look', 'only', 'come', 'its', 'over', 'think', 'also',
  'back', 'after', 'use', 'two', 'how', 'our', 'work', 'first', 'well', 'way',
  'even', 'new', 'want', 'because', 'any', 'these', 'give', 'day', 'most', 'us',
  'great', 'help', 'need', 'house', 'water', 'food', 'please', 'thank', 'thanks',
  'hello', 'hi', 'goodbye', 'bye', 'yes', 'okay', 'sorry', 'excuse',
  'stop', 'start', 'wait', 'here', 'left', 'right',
  'down', 'more', 'less', 'big', 'small', 'hot', 'cold', 'bad',
  'happy', 'sad', 'hungry', 'thirsty', 'tired', 'sick', 'pain', 'hurt',
  'doctor', 'hospital', 'medicine', 'emergency', 'call', 'phone', 'home',
  'family', 'friend', 'mother', 'father', 'brother', 'sister', 'child',
  'love', 'feel', 'eat', 'drink', 'sleep', 'walk',
  'sit', 'stand', 'open', 'close', 'turn', 'light', 'dark', 'room', 'door',
  'bathroom', 'kitchen', 'bed', 'chair', 'table', 'outside', 'inside',
  'morning', 'afternoon', 'evening', 'night', 'today', 'tomorrow', 'yesterday',
  'where', 'why', 'much', 'many', 'again', 'always',
  'never', 'sometimes', 'soon', 'later', 'before',
  'very', 'really', 'too', 'enough', 'still', 'already', 'almost',
  'understand', 'speak', 'listen', 'read', 'write', 'learn', 'teach',
  'name', 'age', 'address', 'number', 'school', 'job', 'money', 'price',
  'beautiful', 'wonderful', 'amazing', 'important', 'different', 'same',
  'together', 'alone', 'quiet', 'loud', 'fast', 'slow', 'easy', 'hard',
];

// Deduplicate while preserving frequency order
const WORD_LIST: string[] = [...new Set(COMMON_WORDS)];

// Common follow-up words (context-aware next-word prediction)
const FOLLOW_UPS: Record<string, string[]> = {
  i:      ['want', 'need', 'am', 'feel', 'can'],
  please: ['help', 'call', 'give', 'open', 'come'],
  can:    ['you', 'i', 'we', 'help', 'go'],
  help:   ['me', 'please', 'now', 'call', 'emergency'],
  want:   ['to', 'food', 'water', 'help', 'go'],
  need:   ['help', 'water', 'food', 'medicine', 'doctor'],
  thank:  ['you', 'thanks', 'so', 'very', 'much'],
  go:     ['home', 'outside', 'to', 'now', 'there'],
  feel:   ['good', 'bad', 'sick', 'tired', 'hungry'],
  am:     ['good', 'hungry', 'tired', 'sick', 'happy'],
};

/**
 * Return up to `limit` word completions for a partial string.
 * Prioritises frequency rank (earlier in list = more common).
 */
function suggestPrefix(partial: string, limit: number = 5): string[] {
  if (!partial || !partial.trim()) return [];

  const prefix = partial.trim().toLowerCase();
  const results: string[] = [];

  for (const word of WORD_LIST) {
    if (word.startsWith(prefix) && word !== prefix) {
      results.push(word);
      if (results.length >= limit) break;
    }
  }

  return results;
}

/**
 * Given the full typed text so far, suggest the next most likely word.
 * Uses prefix matching for incomplete words and context-based follow-ups
 * for completed words.
 */
export function suggestNextWord(context: string, limit: number = 5): string[] {
  if (!context || !context.trim()) {
    // Common conversation starters
    return ['hello', 'i', 'please', 'help', 'thank'].slice(0, limit);
  }

  const words = context.trim().toLowerCase().split(/\s+/);
  const lastWord = words.length > 0 ? words[words.length - 1] : '';

  // If the last word looks incomplete (no trailing space), suggest completions
  if (!context.endsWith(' ')) {
    return suggestPrefix(lastWord, limit);
  }

  // Otherwise suggest common follow-up words
  const followUps = FOLLOW_UPS[lastWord];
  if (followUps) {
    return followUps.slice(0, limit);
  }

  // Default: common words
  return ['please', 'help', 'want', 'need', 'thank'].slice(0, limit);
}

// Debounce utility
let _debounceTimer: ReturnType<typeof setTimeout> | null = null;

/**
 * Debounced version of suggestNextWord.
 * Calls the callback with results after `delayMs` of inactivity.
 */
export function suggestNextWordDebounced(
  context: string,
  callback: (suggestions: string[]) => void,
  limit: number = 5,
  delayMs: number = 150,
): void {
  if (_debounceTimer) clearTimeout(_debounceTimer);
  _debounceTimer = setTimeout(() => {
    callback(suggestNextWord(context, limit));
  }, delayMs);
}
